components/ipfs/ipfs_text_import_worker.cc

```diff
@@ -0,0 +1,97 @@
+
+#include <utility>
+
+#include "base/command_line.h"
+#include "base/files/file_util.h"
+#include "base/guid.h"
+#include "base/task/post_task.h"
+#include "base/task_runner_util.h"
+#include "content/public/browser/browser_context.h"
+#include "net/base/mime_util.h"
+#include "storage/browser/blob/blob_data_builder.h"
+#include "url/gurl.h"
+
+namespace {
+
+const char kIPFSImportMultipartContentType[] = "multipart/form-data;";
+const char kIPFSImportTextMimeType[] = "application/octet-stream";
+const char kFileValueName[] = "file";
+
+using BlobBuilderCallback =
+    base::OnceCallback<std::unique_ptr<storage::BlobDataBuilder>()>;
+
+void AddMultipartHeaderForUploadWithFileName(const std::string& value_name,
+                                             const std::string& file_name,
+                                             const std::string& mime_boundary,
+                                             const std::string& content_type,
+                                             std::string* post_data) {
+  DCHECK(post_data);
+  // First line is the boundary.
+  post_data->append("--" + mime_boundary + "\r\n");
+  // Next line is the Content-disposition.
+  post_data->append("Content-Disposition: form-data; name=\"" + value_name +
+                    "\"; filename=\"" + file_name + "\"\r\n");
+  // If Content-type is specified, the next line is that.
+  post_data->append("Content-Type: " + content_type + "\r\n");
+  // Empty string before next content
+  post_data->append("\r\n");
+}
+
+std::unique_ptr<storage::BlobDataBuilder> BuildBlobWithText(
+    const std::string& text,
+    std::string mime_type,
+    std::string filename,
+    std::string mime_boundary) {
+  auto blob_builder =
+      std::make_unique<storage::BlobDataBuilder>(base::GenerateGUID());
+  std::string post_data_header;
+  AddMultipartHeaderForUploadWithFileName(
+      kFileValueName, filename, mime_boundary, mime_type, &post_data_header);
+  blob_builder->AppendData(post_data_header);
+
+  blob_builder->AppendData(text);
+
+  std::string post_data_footer = "\r\n";
+  net::AddMultipartFinalDelimiterForUpload(mime_boundary, &post_data_footer);
+  blob_builder->AppendData(post_data_footer);
+  return blob_builder;
+}
+
+}  // namespace
+
+namespace ipfs {
+
+IpfsTextImportWorker::IpfsTextImportWorker(content::BrowserContext* context,
+                                           const GURL& endpoint,
+                                           ImportCompletedCallback callback,
+                                           const std::string& text,
+                                           const std::string& host)
+    : IpfsImportWorkerBase(context, endpoint, std::move(callback)) {
+  StartImportText(text, host);
+}
+
+IpfsTextImportWorker::~IpfsTextImportWorker() = default;
+
+void IpfsTextImportWorker::StartImportText(const std::string& text,
+                                           const std::string& host) {
```

Comment:
Should we add some edge case handling for text, host being empty and run callback with failure if so?